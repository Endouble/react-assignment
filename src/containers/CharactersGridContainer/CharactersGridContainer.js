import React, { Component } from 'react';

import CharactersGrid from './../../components/CharactersGrid'

class CharactersGridContainer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      characters: [],
      isLoading: false,
      onError: false
    }
  }

  render() {
    const { onError, characters, isLoading} = this.state;
    return <React.Fragment>
      {isLoading?
        <p>Loading...</p>
        :null
      }
      <CharactersGrid />
      {onError?
        <p> An error ocurred, please try again later... </p>
        :null
      }
    </React.Fragment>
  }
}

export default CharactersGridContainer;