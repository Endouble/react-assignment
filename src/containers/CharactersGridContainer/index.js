import CharactersGridContainer from './CharactersGridContainer';

export default CharactersGridContainer;