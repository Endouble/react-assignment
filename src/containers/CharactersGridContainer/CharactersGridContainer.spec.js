import React from 'react';
import { shallow, mount } from 'enzyme';

import CharactersGridContainer from './CharactersGridContainer';

import swapiMock from './swapi.mocks.json';

const mockResponse = (status, statusText, response) => new window.Response(response, {
    status,
    statusText,
    headers: {
        'Content-type': 'application/json',
    },
});

describe('Characters Grid Container Test', () => {
    let shallowCharactersGridContainer;

    const getShallowCharactersGridContainer = () => {
        if (!shallowCharactersGridContainer) {
            shallowCharactersGridContainer = shallow(<CharactersGridContainer />);
        }
        return shallowCharactersGridContainer;
    };

    beforeEach(() => {
        shallowCharactersGridContainer = undefined;
    });

    test('Component should render without crashing', () => {
        expect(getShallowCharactersGridContainer()).toBeDefined();
    });

    test('Component should render correctly', () => {
        expect(getShallowCharactersGridContainer()).toMatchSnapshot();
    });

    test('Component should call componentDidMount and get characters', () => {
        global.fetch = jest.fn().mockImplementation(
            () => Promise.resolve(mockResponse(200, null, JSON.stringify(swapiMock))),
        );
        const componentDidMount = jest.spyOn(CharactersGridContainer.prototype, 'componentDidMount');
        const wrapper = mount(<CharactersGridContainer />);

        expect(wrapper).toBeDefined();
        expect(componentDidMount).toHaveBeenCalled();
    });

    test('Component should call getStarWarsCharacters when filter is applied', () => {
        const wrapper = getShallowCharactersGridContainer();
        wrapper.instance().getStarWarsCharacters = jest.fn();
        wrapper.instance().filterCharacters({ name: 'Luke Skywalker' });
        expect(wrapper.instance().getStarWarsCharacters).toHaveBeenCalled();
    });

    test('Component should render Characters Grid', () => {
        const wrapper = getShallowCharactersGridContainer();
        expect(wrapper.find('CharactersGrid')).toBeDefined();
    });

    test('Component should show character modal only when character is selected', () => {
        const wrapper = getShallowCharactersGridContainer();
        wrapper.setState({ selectedCharacter: swapiMock.results[0] });
        expect(wrapper.find('CharacterModal')).toBeDefined();

        wrapper.setState({ selectedCharacter: null });
        expect(wrapper.find('CharacterModal')).toHaveLength(0);
    });

    test('Component should show spinner while loading', () => {
        const wrapper = getShallowCharactersGridContainer();
        wrapper.setState({ isLoading: true });
        expect(wrapper.find('Loader')).toBeDefined();
    });

    test('Component should show error message when on error', () => {
        const wrapper = getShallowCharactersGridContainer();
        wrapper.setState({ onError: true });
        expect(wrapper.find('Message')).toBeDefined();
    });
});
