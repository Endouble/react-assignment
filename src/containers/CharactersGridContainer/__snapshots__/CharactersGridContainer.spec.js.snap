// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Characters Grid Container Test Component should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CharactersGridContainer />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <CharactersGrid
          data={null}
          filterCallback={[Function]}
          moreCallback={[Function]}
          showCharacterCallback={[Function]}
/>,
        null,
        <Dimmer
          active={true}
>
          <Loader>
                     Loading 
          </Loader>
</Dimmer>,
        null,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "data": null,
          "filterCallback": [Function],
          "moreCallback": [Function],
          "showCharacterCallback": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      null,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "active": true,
          "children": <Loader>
             Loading 
</Loader>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": " Loading ",
          },
          "ref": null,
          "rendered": " Loading ",
          "type": [Function],
        },
        "type": [Function],
      },
      null,
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <CharactersGrid
            data={null}
            filterCallback={[Function]}
            moreCallback={[Function]}
            showCharacterCallback={[Function]}
/>,
          null,
          <Dimmer
            active={true}
>
            <Loader>
                         Loading 
            </Loader>
</Dimmer>,
          null,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "data": null,
            "filterCallback": [Function],
            "moreCallback": [Function],
            "showCharacterCallback": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "active": true,
            "children": <Loader>
               Loading 
</Loader>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": " Loading ",
            },
            "ref": null,
            "rendered": " Loading ",
            "type": [Function],
          },
          "type": [Function],
        },
        null,
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
